src/Compilers/CSharp/Test/Semantic/Semantics/ForEachTests.cs

```diff
@@ -3689,5 +3689,39 @@ void Goo()
                 //         foreach(var item in nonsenseTuple) {}
                 Diagnostic(ErrorCode.ERR_ForEachMissingMember, "nonsenseTuple").WithArguments("(Nonsense, int)", "GetEnumerator").WithLocation(43, 29));
         }
+
+        [Fact]
+        public void ForEachIterator_RefAssignmentWithoutIdentityConversion()
+        {
+            string source = @"
+using System;
+Span<C2> items = new Span<C2>(new C2[1]);
+foreach (ref C t in items) {}
+class C {}
+class C2 : C {}
+";
+            CreateCompilationWithMscorlibAndSpan(source).VerifyDiagnostics(
+                // (4,21): error CS8173: The expression must be of type 'C' because it is being assigned by reference
+                // foreach (ref C t in items) {}
+                Diagnostic(ErrorCode.ERR_RefAssignmentMustHaveIdentityConversion, "items").WithArguments("C").WithLocation(4, 21)
+            );
+        }
+
+        [Fact]
+        public void ForEachIterator_RefReadonlyAssignmentWithoutIdentityConversion()
+        {
+            string source = @"
+using System;
+Span<C2> items = new Span<C2>(new C2[1]);
```

Comment:
Sorry to belabor this. Please also add a test where we create a `ReadOnlySpan<C2> items` and `foreach` over it.

Reply:
If that test does not have the expected error, we might be able to resolve by simply deleting the check for `builder.CurrentPropertyGetter.RefKind == RefKind.Ref`.

I think we already require the ref kind of `Current { get; }` to be compatible with the ref kind of the iteration variable, so I think we only need to check the iteration variable to know if an identity conversion is required.

Reply:
No need to apologize, I am having a lot of fun working on this stuff. It indeed did not have the expected error. I have removed the suggested statement which indeed fixed it. I added an extra test for that scenario as per your suggestion.

I could not find a test that demonstrated the compatibility check on ref kinds so I have added one extra test for that too. I felt like it adds a new case to the testsuite but if this is considered superfluous I can remove it.

Reply:
Sounds good. We don't worry too much about a test being redundant. Best tip I could give to discover the original would be to disable the code that reports the expected diagnostic and see which existing tests fail.